src/test/java/io/confluent/examples/streams/ReduceTest.java

```diff
@@ -0,0 +1,112 @@
+import org.apache.kafka.streams.kstream.Consumed;
+import org.apache.kafka.streams.kstream.Grouped;
+import org.apache.kafka.streams.kstream.KStream;
+import org.apache.kafka.streams.kstream.KTable;
+import org.apache.kafka.streams.kstream.Produced;
+import org.apache.kafka.test.TestUtils;
+import org.junit.Test;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * How to aggregate messages via `groupByKey()` and `reduce()` to implement concatenation of data.
+ *
+ * This example can be adapted to structured (nested) data formats such as Avro or JSON in case you need to concatenate
+ * only certain field(s) in the input.
+ *
+ * Note: This example uses lambda expressions and thus works with Java 8+ only.
+ *
+ * See {@link ReduceScalaTest} for the equivalent Scala example.
+ */
+public class ReduceTest {
+
+  private static final String inputTopic = "inputTopic";
+  private static final String outputTopic = "outputTopic";
+
+  @Test
+  public void shouldConcatenateWithReduce() {
+    final List<KeyValue<Integer, String>> inputRecords = Arrays.asList(
+        new KeyValue<>(456, "stream"),
+        new KeyValue<>(123, "hello"),
+        new KeyValue<>(123, "world"),
+        new KeyValue<>(456, "all"),
+        new KeyValue<>(123, "kafka"),
+        new KeyValue<>(456, "the"),
+        new KeyValue<>(456, "things"),
+        new KeyValue<>(123, "streams")
+    );
+
+    final Map<Integer, String> expectedOutput = new HashMap<>();
+    expectedOutput.put(456, "stream all the things");
+    expectedOutput.put(123, "hello world kafka streams");
+
+    // Step 1: Create the topology and its configuration
+    final StreamsBuilder builder = createTopology();
+    final Properties streamsConfiguration = createTopologyConfiguration();
+
+    try (final TopologyTestDriver topologyTestDriver = new TopologyTestDriver(builder.build(), streamsConfiguration)) {
+      // Step 2: Write the input
+      IntegrationTestUtils.produceKeyValuesSynchronously(
+          inputTopic, inputRecords, topologyTestDriver, new IntegerSerializer(), new StringSerializer());
+
+      // Step 3: Validate the output
+      final Map<Integer, String> actualOutput = IntegrationTestUtils.drainTableOutput(
+          outputTopic, topologyTestDriver, new IntegerDeserializer(), new StringDeserializer());
+      assertThat(actualOutput).hasSameSizeAs(expectedOutput);
+      assertThat(actualOutput).containsAllEntriesOf(expectedOutput);
+    }
+  }
+
+  private StreamsBuilder createTopology() {
+    final StreamsBuilder builder = new StreamsBuilder();
+    final KStream<Integer, String> input = builder.stream(inputTopic, Consumed.with(Serdes.Integer(), Serdes.String()));
+    final KTable<Integer, String> concatenated =
+        input.groupByKey(Grouped.with(Serdes.Integer(), Serdes.String())).reduce((v1, v2) -> v1 + " " + v2);
```

Comment:
Similar to the other example, do you mind giving a textual description of the app? It's just a little distracting to have to figure out what it's doing, when the point is really how to test it.

Reply:
Ack